src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2796,52 +2809,130 @@ private ImmutableArray<Diagnostic> GetDiagnosticsForMethodBodiesInTree(SyntaxTre
-            //                                  we already compiled and not recompile them. It might be also worth checking if any
-            //                                  global using in the tree is still considered unused.
-            SyntaxTree? filterTree = tree;
-            TextSpan? filterSpan = span;
-
-            if (reportUnusedUsings)
+            if (reportUnusedUsings && UsageOfUsingsRecordedInTrees is not null)
             {
                 foreach (var singleDeclaration in ((SourceNamespaceSymbol)SourceModule.GlobalNamespace).MergedDeclaration.Declarations)
                 {
                     if (singleDeclaration.SyntaxReference.SyntaxTree == tree)
                     {
                         if (singleDeclaration.HasGlobalUsings)
                         {
-                            // PROTOTYPE(GlobalUsingDirective): Compile the world for now.
-                            filterTree = null;
-                            filterSpan = null;
+                            // Global Using directives can be used in any tree. Make sure we collect usage information from all of them.
+                            recordUsageOfUsingsInAllTrees = true;
                         }
 
                         break;
                     }
                 }
             }
 
-            MethodCompiler.CompileMethodBodies(
-                compilation: this,
-                moduleBeingBuiltOpt: null,
-                emittingPdb: false,
-                emitTestCoverageData: false,
-                hasDeclarationErrors: false,
-                emitMethodBodies: false,
-                diagnostics: bindingDiagnostics,
-                filterOpt: filterTree is object ? (Predicate<Symbol>?)(s => IsDefinedOrImplementedInSourceTree(s, filterTree, filterSpan)) : (Predicate<Symbol>?)null,
-                cancellationToken: cancellationToken);
+            if (recordUsageOfUsingsInAllTrees && UsageOfUsingsRecordedInTrees?.IsEmpty == true)
+            {
+                Debug.Assert(reportUnusedUsings);
 
-            DocumentationCommentCompiler.WriteDocumentationCommentXml(this, null, null, bindingDiagnostics, cancellationToken, filterTree, filterSpan);
+                // Simply compile the world
+                compileMethodBodiesAndDocComments(filterTree: null, filterSpan: null, bindingDiagnostics, cancellationToken);
+                _usageOfUsingsRecordedInTrees = null;
+            }
+            else
+            {
+                // Always compile the target tree
+                compileMethodBodiesAndDocComments(filterTree: tree, filterSpan: span, bindingDiagnostics, cancellationToken);
+
+                if (reportUnusedUsings)
+                {
+                    registeredUsageOfUsingsInTree(tree);
+                }
+
+                // Compile other trees if we need to, but discard diagnostics from them.
+                if (recordUsageOfUsingsInAllTrees)
```

Comment:
>recordUsageOfUsingsInAllTrees [](start = 20, length = 29)

Does this mean we're compiling all trees in the compilation twice if there are any global usings in the compilation?

Reply:
> Does this mean we're compiling all trees in the compilation twice if there are any global usings in the compilation?

This function is not called during normal build process. It is only called when a client requests diagnostics for a specific syntax tree through SemanticModel. That tree wiil always be compiled. When it contains global usings, we will also compile other trees, unless we know that we already did that for them.  

---
In reply to: [602637179](https://github.com/dotnet/roslyn/pull/52132#discussion_r602637179) [](ancestors = 602637179)